test/ref_ops_test.cpp

```diff
@@ -2439,6 +2439,100 @@ TEST_CASE(logsoftmax_test_axis_3)
     EXPECT(migraphx::verify_range(results_vector, s));
 }
 
+TEST_CASE(loop_test)
+{
+    migraphx::shape si{migraphx::shape::int64_type};
+    migraphx::shape s{migraphx::shape::int64_type, {1}};
+    migraphx::shape sc{migraphx::shape::bool_type};
+    auto create_program = [&]() {
+        migraphx::program p;
+        auto* mm = p.get_main_module();
+
+        auto in_iter = mm->add_parameter("iter_num", si);
+        auto in_cond = mm->add_parameter("ccond", sc);
+        auto in_val  = mm->add_parameter("val", s);
+
+        auto* body = p.create_module("loop_module");
+        auto iter  = body->add_parameter("#loop_module_in_0", si);
+        body->add_parameter("#loop_module_in_1", sc);
+        auto in_v               = body->add_parameter("#loop_module_in_2", s);
+        std::vector<int64_t> vd = {3};
+        auto l                  = body->add_literal(migraphx::literal(si, vd));
+        auto ad                 = body->add_instruction(migraphx::make_op("add"), iter, l);
+        auto val                = body->add_instruction(migraphx::make_op("add"), in_v, ad);
+        auto eq                 = body->add_instruction(migraphx::make_op("equal"), iter, l);
+        auto beq                = body->add_instruction(
+            migraphx::make_op("convert", {{"target_type", migraphx::shape::bool_type}}), eq);
+        auto neq = body->add_instruction(migraphx::make_op("not"), beq);
+        body->add_return({neq, val, val});
+
+        auto rl = mm->add_instruction(migraphx::make_op("loop", {{"max_iterations", 10}}),
+                                      {in_iter, in_cond, in_val},
+                                      {body});
+        auto r0 = mm->add_instruction(migraphx::make_op("get_tuple_elem", {{"index", 0}}), rl);
+        auto r1 = mm->add_instruction(migraphx::make_op("get_tuple_elem", {{"index", 1}}), rl);
+        mm->add_return({r0, r1});
+
+        return p;
+    };
+
+    auto run_prog = [&](int64_t iter_num, bool cond, int64_t val) {
+        auto p = create_program();
+        p.compile(migraphx::ref::target{});
+        migraphx::parameter_map pp;
+        pp["iter_num"] = migraphx::argument(si, &iter_num);
+        pp["ccond"]    = migraphx::argument(sc, &cond);
+        pp["val"]      = migraphx::argument(s, &val);
+        auto rets      = p.eval(pp);
+
+        std::vector<std::vector<int64_t>> res;
+        for(auto& arg : rets)
+        {
+            std::vector<int64_t> vec;
+            arg.visit([&](auto v) { vec.assign(v.begin(), v.end()); });
+            res.push_back(vec);
+        }
+
+        return res;
+    };
```

Comment:
I think it would be better to make this a regular function and make all the test cases seperate `TEST_CASE` cases in the test suite. This will make it easier to debug in the future as we can just run one test case(instead of all of them).

Reply:
changed